File: backend/routes/investigation.py
"""
IP Investigation Routes - Deep analysis of specific IP addresses

Provides comprehensive reporting on IP behavior, connections, and security analysis.
"""

from flask import Blueprint, jsonify, request
import sqlite3
from datetime import datetime
from typing import Dict, Any, List

import sys
import os
sys.path.append(os.path.dirname(os.path.dirname(__file__)))
from utils.db import get_db_connection

investigation_bp = Blueprint('investigation', __name__)


def _pcap_exists(conn, pcap_id: int) -> bool:
    """Check if PCAP exists."""
    cur = conn.cursor()
    cur.execute('SELECT 1 FROM pcaps WHERE id = ?', (pcap_id,))
    return cur.fetchone() is not None


@investigation_bp.route('/investigation/<int:pcap_id>/<ip_address>', methods=['GET'])
def investigate_ip(pcap_id: int, ip_address: str):
    """
    Get comprehensive investigation report for a specific IP address.
    
    Returns:
    - IP overview (MAC, first/last seen, duration)
    - Traffic statistics (packets, bytes, protocols)
    - Domains accessed by this IP
    - Connection flows (incoming/outgoing)
    - Security analysis (with Gemini AI)
    """
    try:
        with get_db_connection() as conn:
            if not _pcap_exists(conn, pcap_id):
                return jsonify({"error": "PCAP not found"}), 404
            
            cur = conn.cursor()
            
            # 1. Get IP Overview from devices table
            device_row = cur.execute(
                'SELECT ip, mac, hostname, first_seen, last_seen FROM devices WHERE pcap_id = ? AND ip = ?',
                (pcap_id, ip_address)
            ).fetchone()
            
            if not device_row:
                return jsonify({"error": "IP address not found in this PCAP"}), 404
            
            ip, mac, hostname, first_seen, last_seen = device_row
            duration_sec = (last_seen - first_seen) if (first_seen and last_seen) else 0
            
            # 2. Get Traffic Statistics from ip_observations
            ip_obs_row = cur.execute(
                'SELECT count FROM ip_observations WHERE pcap_id = ? AND ip = ?',
                (pcap_id, ip_address)
            ).fetchone()
            
            total_packets = ip_obs_row[0] if ip_obs_row else 0
            
            # 3. Get Flow Statistics (outgoing and incoming)
            # Outgoing flows (this IP as source)
            outgoing_flows = cur.execute(
                'SELECT dst_ip, dst_port, protocol, packet_count, byte_count, first_seen, last_seen '
                'FROM flows WHERE pcap_id = ? AND src_ip = ? ORDER BY byte_count DESC LIMIT 50',
                (pcap_id, ip_address)
            ).fetchall()
            
            # Incoming flows (this IP as destination)
            incoming_flows = cur.execute(
                'SELECT src_ip, src_port, protocol, packet_count, byte_count, first_seen, last_seen '
                'FROM flows WHERE pcap_id = ? AND dst_ip = ? ORDER BY byte_count DESC LIMIT 50',
                (pcap_id, ip_address)
            ).fetchall()
            
            # Calculate traffic stats
            outgoing_packets = sum(f[3] for f in outgoing_flows)
            outgoing_bytes = sum(f[4] for f in outgoing_flows)
            incoming_packets = sum(f[3] for f in incoming_flows)
            incoming_bytes = sum(f[4] for f in incoming_flows)
            
            # Protocol breakdown
            protocols = {}
            for flow in outgoing_flows:
                proto = flow[2]
                protocols[proto] = protocols.get(proto, 0) + flow[3]
            
            # 4. Get Domains accessed by this IP
            domains = cur.execute(
                'SELECT domain, source, count, verdict, explanation '
                'FROM domains WHERE pcap_id = ? AND ip = ? ORDER BY count DESC',
                (pcap_id, ip_address)
            ).fetchall()
            
            # Count verdicts
            safe_count = sum(1 for d in domains if d[3] == 'safe')
            suspicious_count = sum(1 for d in domains if d[3] == 'suspicious')
            malicious_count = sum(1 for d in domains if d[3] == 'malicious')
            unknown_count = sum(1 for d in domains if not d[3] or d[3] == 'unknown')
            
            # Get suspicious/malicious domain details
            risky_domains = [
                {'domain': d[0], 'verdict': d[3], 'explanation': d[4], 'count': d[2]}
                for d in domains if d[3] in ['suspicious', 'malicious']
            ]
            
            # 5. Top connections
            top_connections = []
            for flow in outgoing_flows[:10]:  # Top 10
                dst_ip, dst_port, protocol, pkt_count, byte_count, f_first, f_last = flow
                duration = (f_last - f_first) if (f_first and f_last) else 0
                top_connections.append({
                    'destination': dst_ip,
                    'port': dst_port,
                    'protocol': protocol,
                    'packets': pkt_count,
                    'bytes': byte_count,
                    'duration_sec': duration
                })
            
            # 6. Build comprehensive report
            report = {
                'success': True,
                'pcap_id': pcap_id,
                'ip_address': ip_address,
                
                # Overview
                'overview': {
                    'ip': ip,
                    'mac': mac,
                    'hostname': hostname,
                    'first_seen': first_seen,
                    'last_seen': last_seen,
                    'duration_sec': duration_sec,
                    'total_packets': total_packets
                },
                
                # Traffic Statistics
                'traffic': {
                    'outgoing': {
                        'packets': outgoing_packets,
                        'bytes': outgoing_bytes
                    },
                    'incoming': {
                        'packets': incoming_packets,
                        'bytes': incoming_bytes
                    },
                    'total': {
                        'packets': outgoing_packets + incoming_packets,
                        'bytes': outgoing_bytes + incoming_bytes
                    },
                    'protocols': protocols
                },
                
                # Domains
                'domains': {
                    'total': len(domains),
                    'safe': safe_count,
                    'suspicious': suspicious_count,
                    'malicious': malicious_count,
                    'unknown': unknown_count,
                    'risky_domains': risky_domains,
                    'all_domains': [
                        {'domain': d[0], 'count': d[2], 'verdict': d[3], 'explanation': d[4]}
                        for d in domains
                    ]
                },
                
                # Connections
                'connections': {
                    'outgoing_count': len(outgoing_flows),
                    'incoming_count': len(incoming_flows),
                    'top_connections': top_connections
                }
            }
            
            # 7. Generate AI Analysis using Gemini
            try:
                from services.ip_investigator import analyze_ip_behavior
                ai_analysis = analyze_ip_behavior(report)
                report['ai_analysis'] = ai_analysis
            except Exception as e:
                print(f"Gemini analysis failed: {e}")
                report['ai_analysis'] = {
                    'error': 'AI analysis unavailable',
                    'risk_level': 'unknown'
                }
            
            return jsonify(report), 200
            
    except Exception as e:
        import traceback
        print(f"Investigation error: {e}")
        print(traceback.format_exc())
        return jsonify({"error": f"Investigation failed: {str(e)}"}), 500


@investigation_bp.route('/investigation/<int:pcap_id>/ips', methods=['GET'])
def list_ips_for_investigation(pcap_id: int):
    """
    Get list of all IPs in a PCAP for investigation selection.
    """
    try:
        with get_db_connection() as conn:
            if not _pcap_exists(conn, pcap_id):
                return jsonify({"error": "PCAP not found"}), 404
            
            cur = conn.cursor()
            rows = cur.execute(
                'SELECT ip, count FROM ip_observations WHERE pcap_id = ? ORDER BY count DESC',
                (pcap_id,)
            ).fetchall()
            
            ips = [{'ip': r[0], 'packet_count': r[1]} for r in rows]
            
            return jsonify({
                'success': True,
                'pcap_id': pcap_id,
                'total': len(ips),
                'ips': ips
            }), 200
            
    except Exception as e:
        print(f"List IPs error: {e}")
        return jsonify({"error": f"Failed to list IPs: {str(e)}"}), 500

File: electron-app/src/js/investigation.js
// IP Investigation page logic

(function () {
  const state = {
    fileId: null,
    currentIP: null,
    reportData: null
  };

  function getSelectedFileId() {
    const id = sessionStorage.getItem('selectedFileId');
    return id ? parseInt(id, 10) : null;
  }

  function formatBytes(bytes) {
    if (!bytes && bytes !== 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`;
  }

  function formatTimestamp(ts) {
    if (!ts) return 'N/A';
    const d = new Date(ts * 1000);
    return d.toLocaleString();
  }

  function formatDuration(seconds) {
    if (!seconds) return '0s';
    const h = Math.floor(seconds / 3600);
    const m = Math.floor((seconds % 3600) / 60);
    const s = Math.floor(seconds % 60);
    if (h > 0) return `${h}h ${m}m ${s}s`;
    if (m > 0) return `${m}m ${s}s`;
    return `${s}s`;
  }

  function getRiskColor(riskLevel) {
    const colors = {
      'low': '#28a745',
      'medium': '#ffc107',
      'high': '#fd7e14',
      'critical': '#dc3545',
      'unknown': '#6c757d'
    };
    return colors[riskLevel] || colors.unknown;
  }

  async function loadAvailableIPs() {
    try {
      const data = await window.api.getInvestigationIPs(state.fileId);
      const ipList = document.getElementById('ipList');
      
      if (!data.ips || data.ips.length === 0) {
        ipList.innerHTML = '<span class="text-muted">No IPs found</span>';
        return;
      }

      // Show top 10 IPs by packet count
      const topIPs = data.ips.slice(0, 10);
      ipList.innerHTML = topIPs.map(item => 
        `<button class="ip-chip" data-ip="${item.ip}">${item.ip} (${item.packet_count})</button>`
      ).join('');

      // Add click handlers
      document.querySelectorAll('.ip-chip').forEach(chip => {
        chip.onclick = () => {
          const ip = chip.getAttribute('data-ip');
          document.getElementById('ipInput').value = ip;
          investigateIP(ip);
        };
      });
    } catch (e) {
      console.error('Failed to load IPs:', e);
      window.ui.showStatus('Failed to load IP list', 'error');
    }
  }

  async function investigateIP(ip) {
    if (!ip) {
      window.ui.showStatus('Please enter an IP address', 'warning');
      return;
    }

    // Show loading
    document.getElementById('loadingIndicator').style.display = 'block';
    document.getElementById('reportSection').style.display = 'none';

    try {
      const data = await window.api.investigateIP(state.fileId, ip);
      
      if (!data.success) {
        throw new Error(data.error || 'Investigation failed');
      }

      state.currentIP = ip;
      state.reportData = data;

      renderReport(data);

      // Hide loading, show report
      document.getElementById('loadingIndicator').style.display = 'none';
      document.getElementById('reportSection').style.display = 'block';

      // Load threat intelligence
      loadThreatIntelligence(ip);

      window.ui.showStatus(`Investigation complete for ${ip}`, 'success');

    } catch (e) {
      console.error('Investigation error:', e);
      document.getElementById('loadingIndicator').style.display = 'none';
      window.ui.showStatus(`Investigation failed: ${e.message}`, 'error');
    }
  }

  function renderReport(data) {
    // 1. Overview
    const overview = data.overview || {};
    document.getElementById('overviewIP').textContent = overview.ip || 'N/A';
    document.getElementById('overviewMAC').textContent = overview.mac || 'N/A';
    document.getElementById('overviewFirst').textContent = formatTimestamp(overview.first_seen);
    document.getElementById('overviewLast').textContent = formatTimestamp(overview.last_seen);
    document.getElementById('overviewDuration').textContent = formatDuration(overview.duration_sec);
    document.getElementById('overviewPackets').textContent = (overview.total_packets || 0).toLocaleString();

    // 2. AI Analysis
    const ai = data.ai_analysis || {};
    const riskLevel = (ai.risk_level || 'unknown').toUpperCase();
    const riskScore = ai.risk_score || 0;
    
    const riskBadge = document.getElementById('riskBadge');
    riskBadge.textContent = riskLevel;
    riskBadge.style.backgroundColor = getRiskColor(ai.risk_level);
    
    document.getElementById('aiSummary').innerHTML = `<p>${ai.summary || 'No summary available'}</p>`;
    
    const scoreFill = document.getElementById('scoreFill');
    scoreFill.style.width = `${riskScore}%`;
    scoreFill.style.backgroundColor = getRiskColor(ai.risk_level);
    document.getElementById('scoreValue').textContent = `${riskScore}/100`;

    // 3. Traffic Statistics
    const traffic = data.traffic || {};
    const outgoing = traffic.outgoing || {};
    const incoming = traffic.incoming || {};
    const total = traffic.total || {};

    document.getElementById('trafficOutgoing').textContent = formatBytes(outgoing.bytes);
    document.getElementById('trafficOutgoingPackets').textContent = `${(outgoing.packets || 0).toLocaleString()} packets`;
    
    document.getElementById('trafficIncoming').textContent = formatBytes(incoming.bytes);
    document.getElementById('trafficIncomingPackets').textContent = `${(incoming.packets || 0).toLocaleString()} packets`;
    
    document.getElementById('trafficTotal').textContent = formatBytes(total.bytes);
    document.getElementById('trafficTotalPackets').textContent = `${(total.packets || 0).toLocaleString()} packets`;

    // Protocol breakdown
    const protocols = traffic.protocols || {};
    const protocolList = document.getElementById('protocolList');
    if (Object.keys(protocols).length > 0) {
      protocolList.innerHTML = Object.entries(protocols)
        .map(([proto, count]) => `
          <div class="protocol-item">
            <span class="protocol-name">${proto}</span>
            <span class="protocol-count">${count.toLocaleString()}</span>
          </div>
        `).join('');
    } else {
      protocolList.innerHTML = '<p class="text-muted">No protocol data</p>';
    }

    // 4. Domains
    const domains = data.domains || {};
    document.getElementById('domainTotal').textContent = domains.total || 0;
    document.getElementById('domainSafe').textContent = domains.safe || 0;
    document.getElementById('domainSuspicious').textContent = domains.suspicious || 0;
    document.getElementById('domainMalicious').textContent = domains.malicious || 0;
    document.getElementById('domainUnknown').textContent = domains.unknown || 0;

    // Risky domains
    const riskyDomains = document.getElementById('riskyDomains');
    const risky = domains.risky_domains || [];
    if (risky.length > 0) {
      riskyDomains.style.display = 'block';
      riskyDomains.innerHTML = '<h4>⚠️ Domains Requiring Attention:</h4>' + risky.map(d => `
        <div class="risky-domain-item ${d.verdict}">
          <div class="domain-name">
            <i class="fa-solid ${d.verdict === 'malicious' ? 'fa-skull-crossbones' : 'fa-exclamation-triangle'}"></i>
            ${d.domain}
          </div>
          <div class="domain-verdict">${d.verdict.toUpperCase()}</div>
          <div class="domain-explanation">${d.explanation || 'No explanation'}</div>
          <div class="domain-count">Queried ${d.count} times</div>
        </div>
      `).join('');
    } else {
      riskyDomains.style.display = 'none';
    }

    // All domains
    const allDomains = domains.all_domains || [];
    document.getElementById('allDomainsCount').textContent = allDomains.length;
    const allDomainsList = document.getElementById('allDomainsList');
    if (allDomains.length > 0) {
      allDomainsList.innerHTML = allDomains.map(d => `
        <div class="domain-item">
          <span class="domain-name">${d.domain}</span>
          <span class="verdict-badge verdict-${d.verdict || 'unknown'}">${(d.verdict || 'unknown').toUpperCase()}</span>
          <span class="domain-count">${d.count} queries</span>
        </div>
      `).join('');
    } else {
      allDomainsList.innerHTML = '<p class="text-muted">No domains accessed</p>';
    }

    // 5. Findings
    const findings = ai.findings || [];
    const findingsList = document.getElementById('findingsList');
    if (findings.length > 0) {
      findingsList.innerHTML = findings.map(f => `<li>${f}</li>`).join('');
    } else {
      findingsList.innerHTML = '<li class="text-muted">No specific findings</li>';
    }

    // 6. Recommendations
    const recommendations = ai.recommendations || [];
    const recList = document.getElementById('recommendationsList');
    if (recommendations.length > 0) {
      recList.innerHTML = recommendations.map(r => `<li>${r}</li>`).join('');
    } else {
      recList.innerHTML = '<li class="text-muted">No recommendations</li>';
    }

    // 7. Top Connections
    const connections = data.connections || {};
    const topConns = connections.top_connections || [];
    const connsTable = document.getElementById('connectionsTable');
    if (topConns.length > 0) {
      connsTable.innerHTML = topConns.map(c => `
        <tr>
          <td>${c.destination}</td>
          <td>${c.port || 'N/A'}</td>
          <td>${c.protocol}</td>
          <td>${(c.packets || 0).toLocaleString()}</td>
          <td>${formatBytes(c.bytes)}</td>
          <td>${formatDuration(c.duration_sec)}</td>
        </tr>
      `).join('');
    } else {
      connsTable.innerHTML = '<tr><td colspan="6" class="text-center text-muted">No connections</td></tr>';
    }
  }

  async function loadThreatIntelligence(ip) {
    const card = document.getElementById('threatIntelCard');
    const loading = document.getElementById('threatIntelLoading');
    const content = document.getElementById('threatIntelContent');
    const noData = document.getElementById('noThreatIntelData');

    // Show the card and loading state
    card.style.display = 'block';
    loading.style.display = 'flex';
    content.style.display = 'none';

    try {
      const data = await window.api.checkIPThreat(ip);

      // Hide loading
      loading.style.display = 'none';
      content.style.display = 'block';

      // Check if any services returned data
      const sources = data.sources || [];
      if (sources.length === 0) {
        noData.style.display = 'block';
        return;
      }

      noData.style.display = 'none';

      // Overall threat score
      const threatScore = data.threat_score || 0;
      const threatLevel = (data.threat_level || 'clean').toUpperCase();
      
      document.getElementById('threatScoreNumber').textContent = Math.round(threatScore);
      document.getElementById('threatLevelBadge').textContent = threatLevel;
      document.getElementById('threatLevelBadge').className = `threat-level-badge threat-${data.threat_level}`;

      // Color the score circle
      const scoreCircle = document.getElementById('threatScoreCircle');
      if (threatScore >= 80) {
        scoreCircle.className = 'score-circle critical';
      } else if (threatScore >= 60) {
        scoreCircle.className = 'score-circle high';
      } else if (threatScore >= 40) {
        scoreCircle.className = 'score-circle medium';
      } else if (threatScore >= 20) {
        scoreCircle.className = 'score-circle low';
      } else {
        scoreCircle.className = 'score-circle clean';
      }

      // Summary
      if (data.is_malicious) {
        document.getElementById('threatSummary').innerHTML = 
          '<i class="fa-solid fa-exclamation-triangle"></i> This IP has been flagged as malicious by threat intelligence sources';
      } else if (threatScore >= 40) {
        document.getElementById('threatSummary').innerHTML = 
          '<i class="fa-solid fa-exclamation-circle"></i> This IP shows suspicious activity patterns';
      } else {
        document.getElementById('threatSummary').textContent = 
          'No significant threats detected from this IP';
      }

      // VirusTotal data
      const vtSource = sources.find(s => s.source === 'VirusTotal');
      if (vtSource && !vtSource.error) {
        document.getElementById('virusTotalSection').style.display = 'block';
        document.getElementById('vtBadge').textContent = `${Math.round(vtSource.score)}%`;
        document.getElementById('vtBadge').className = 'source-badge ' + 
          (vtSource.score >= 50 ? 'badge-danger' : vtSource.score >= 20 ? 'badge-warning' : 'badge-success');
        
        document.getElementById('vtRatio').textContent = 
          `${vtSource.malicious_count}/${vtSource.total_engines}`;
        document.getElementById('vtMalicious').textContent = vtSource.malicious_count || 0;
        document.getElementById('vtSuspicious').textContent = vtSource.suspicious_count || 0;
        document.getElementById('vtASOwner').textContent = vtSource.as_owner || 'Unknown';
      }

      // AbuseIPDB data
      const abuseSource = sources.find(s => s.source === 'AbuseIPDB');
      if (abuseSource && !abuseSource.error) {
        document.getElementById('abuseIPDBSection').style.display = 'block';
        document.getElementById('abuseIPDBBadge').textContent = `${Math.round(abuseSource.score)}%`;
        document.getElementById('abuseIPDBBadge').className = 'source-badge ' + 
          (abuseSource.score >= 50 ? 'badge-danger' : abuseSource.score >= 20 ? 'badge-warning' : 'badge-success');
        
        document.getElementById('abuseConfidence').textContent = `${abuseSource.abuse_confidence_score}%`;
        document.getElementById('abuseReports').textContent = abuseSource.total_reports || 0;
        document.getElementById('abuseLastReported').textContent = 
          abuseSource.last_reported ? new Date(abuseSource.last_reported).toLocaleDateString() : 'Never';
        document.getElementById('abuseISP').textContent = abuseSource.isp || 'Unknown';
      }

    } catch (error) {
      console.error('Threat intelligence error:', error);
      loading.style.display = 'none';
      content.style.display = 'block';
      noData.style.display = 'block';
    }
  }

  async function init() {
    state.fileId = getSelectedFileId();
    
    if (!state.fileId) {
      window.ui.showStatus('No file selected. Please go to Files and analyze a PCAP first.', 'warning');
      document.querySelector('.search-section').style.opacity = '0.5';
      document.querySelector('.search-section').style.pointerEvents = 'none';
      return;
    }

    // Load available IPs
    await loadAvailableIPs();

    // Wire up search button
    document.getElementById('investigateBtn').onclick = () => {
      const ip = document.getElementById('ipInput').value.trim();
      investigateIP(ip);
    };

    // Enter key support
    document.getElementById('ipInput').onkeypress = (e) => {
      if (e.key === 'Enter') {
        const ip = document.getElementById('ipInput').value.trim();
        investigateIP(ip);
      }
    };

    // Threat intel refresh button
    const refreshBtn = document.getElementById('refreshThreatIntel');
    if (refreshBtn) {
      refreshBtn.onclick = () => {
        if (state.currentIP) {
          loadThreatIntelligence(state.currentIP);
        }
      };
    }
  }

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', init);
  } else {
    init();
  }
})();
